File: FindKClosestElementsBinarySearch.java
// TC: O(log(n - k))
// SC: O(1)
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        List<Integer> res = new ArrayList<>();
        if(arr == null || arr.length == 0) {
            return res;
        }
        // Using binary search to find the 1st element of the resultant k elements
        int low, high, mid;
        low = 0;
        high = arr.length - k;
        while(low < high) {
            mid = low + (high - low) / 2;

            int dFirst = x - arr[mid];
            int dLast = arr[mid + k] - x;

            if(dFirst > dLast) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        for(int i = low; i < low + k; i++) {
            res.add(arr[i]);
        }
        return res;
    }
}

File: find_k_closest_elements_binary_search.py
# Time Complexity : O(log(n-k)) + O(k) where n length of the array and k is number closest elements to findout
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        low = 0
        high = len(arr) - k
        
        while low < high:
            mid = low + (high - low) // 2
            diffL = x - arr[mid]
            diffH = arr[mid + k] - x
            
            if diffL > diffH:
                low = mid + 1
            else:
                high = mid
            
        return arr[low: low + k]